python/lsst/obs/base/exposureAssembler.py

```diff
@@ -237,12 +229,19 @@ def assemble(self, components):
         info.setDetector(components.pop("detector", None))
         info.setTransmissionCurve(components.pop("transmissionCurve", None))
 
-        # Filter needs to be updated specially to match Exposure behavior
+        # A FilterLabel component gets read correctly even for version 1
+        # Exposure files, which persist the info as a Filter.
+        # TODO: deprecate in DM-27177, remove in DM-27811
+        info.setFilterLabel(components.pop("filterLabel", None))
+
+        # Filter needs to be updated specially to match Exposure v1 behavior
         # from ExposureFitsReader::MetadataReader
+        # TODO: deprecate in DM-27170, remove in DM-27177
 
         # Override the serialized filter knowing that we are using FILTER
         md = info.getMetadata()
-        if "filter" in components and "FILTER" in md:
+        if "filter" in components and "FILTER" in md \
```

Comment:
I don't think anyone has ever used composite disassembly in a supported repo so I don't think you will ever need to worry about a filter component from a disassembled Exposure being persisted anywhere. In fact, I think it would be best if we made it so that disassembly never returned a `filter=afwFilter` so that one should never turn up in the future (assuming that if people put an Exposure in the future they will always have a filterLabel).

Reply:
> I think it would be best if we made it so that disassembly never returned a filter=afwFilter so that one should never turn up in the future

This would mean introducing an inconsistency between the APIs for disassembled composites, non-disassembled composites, and the `Exposure` class itself. I think that would make the code very difficult for users to work with. Or are you suggesting that `.filter` persist as `FilterLabel` but convert to `Filter` on read?

Reply:
I'm specifically talking about disassembly and assembly. You disassemble into components and then you put it all back together again. This is distinct from asking for individual components. In this case if you have both filter and filterLabel components then these are duplicating information and you only need one of them to recreate the Exposure in assembly. I don't understand why you need both and why disassembly can't ignore `filter` and only return a `filterLabel`.

Reply:
Just to make sure I understand this: if the Butler has a disassembled exposure, and the user calls `get('exposure.filter')`, are you saying that this can (partially?) reassemble the exposure from its `filterLabel`, then use different code to get the filter (e.g., by calling `Exposure.getFilter()` on the reassembled composite)?

Reply:
Sorry, but after some experimentation it looks like any other approach would require some serious hacking around the Butler's built-in behavior. As far as I can tell from the various checks I'm triggering, it really wasn't designed to support a distinction between component as something requested by the user and component as something stored on disk.

Reply:
What I'm trying to say is that you can make `filter` a derived component and `filterLabel` be the proper component that is disassembled. This is exactly the infrastructure to use to support a distinction between something requested by the user and something stored on disk. Then if someone asks for a filter it's like asking for the bbox. The system declares that to derive filter it needs the real filterLabel component.

Reply:
Ok, I think I see what you mean -- you're referring to the `derivedComponents` block in `storageClasses.yaml`. I didn't try that, because I didn't know what it was for.

Reply:
Yes. Exactly that. derivedComponents. We shouldn't be storing the same information in two different components but if we can derive the second component from the first then that is the right thing to do.

Reply:
And do I now need a `FilterLabelAssembler` to do the translation? Do I only need to override `getComponent`? Butler is giving me an error about a missing delegate for the `FilterLabel` storage class, but the `StorageClassDelegate` documentation doesn't explicitly say anything about being used for derived components.

Reply:
For information on delegates see https://pipelines.lsst.io/v/weekly/modules/lsst.daf.butler/formatters.html#storage-class-delegates

Reply:
Actually, after reading that I'm more confused. The docs seem to say that I should not declare a delegate for `FilterLabel` (which itself has no components) and should support the translation in `ExposureAssembler.getComponent`. However, that method already supports "filter" through `EXPOSURE_READ_COMPONENTS` and the default getter-finder; this also seems to be how `bbox` and `dimensions` are supported.

For reference, the error I'm getting is 
```
E                   ValueError: Failure from formatter 'lsst.obs.base.formatters.fitsGeneric.FitsGenericFormatter' for dataset 5 
(unknown.filter from 
file:///scratch/krzys001/obs_base/tests/tmp22tb_tqn/sub%3F%23dir/test297675710/unknown.filterLabel/d-r/42/unknown_filterLabel_d-r_42_DummyCam_test297675710.fits):
No delegate class is associated with StorageClass FilterLabel
```

Reply:
Is the rule really "any storage class associated with a component returned by `selectResponsibleComponent` needs to have a delegate"?

Reply:
Sorry. You are the first person that's tried to do this other than me. Filter is not a real component any more -- the delegate is there to help a `filter` request when a composite is disassembled such that the datastore knows that to calculate `filter` it needs to read in `filterLabel`.

Reply:
Finally got all unit tests passing for this. Thank you for your help, and your patience. Because it's a significant rewrite, I'd like to put this branch back into review once I confirm that `ci_hsc` can't break it.

Reply:
I think this might be an example of why we need some documentation in those butler YAML files, a la DM-23603.